### PR TITLE
File manager

### PR DESCRIPTION
**Pull request by gongweibao**
功能里边实现了server的ls功能。

### CODE REVIEW

File: go/file_manager/pfsserver/handler.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+import (
+	"encoding/json"
+	"github.com/cloud/go/file_manager/pfscommon"
+	"github.com/cloud/go/file_manager/pfsmodules"
+	"log"
+	"net/http"
+)
+
+func GetFiles(w http.ResponseWriter, r *http.Request) {
```
Comment (typhoonzero) [inline go/file_manager/pfsserver/handler.go]: Handlers should be of the name like `GetFilesHandler`

File: go/file_manager/pfscommon/parsebody.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+	"log"
+	"net/http"
+)
+
+type Rep interface {
+	GetErr() string
+	SetErr(err string)
+}
+
+func MakeResponse(w http.ResponseWriter,
```
Comment (typhoonzero) [inline go/file_manager/pfscommon/parsebody.go]: API responses can be generalized like:
```go

type Response interface {

	// WriteJsonResponse should be implemeted at child Response
	WriteJsonResponse(w *http.ResponseWriter) error
	// general writer to call
	writeResponse(w *http.ResponseWriter)
}

type BaseResponse struct {
	returnCode uint32
	errMsg     string
	headers    map[string]string
}

// WriteJsonResponse for BaseResponse is abstract
func (r *BaseResponse) WriteJsonResponse(w *http.ResponseWriter) error {
	return nil
}

type FilesResponse struct {
	BaseResponse
	files []string
}

// Override
func (r *FilesResponse) WriteJsonResponse(w *http.ResponseWriter) error {
}

func MyHandler(w http.ResponseWriter, r *http.Request) {
	resp := FilesResponse{}
	resp.WriteJsonResponse()
}
```

File: go/file_manager/pfscommon/checkcmd.go [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+type Cmd struct {
+	Method  string
+	Options []string
+	Src     []FilePath
+	Dest    FilePath
+}
+
+func CheckLsSynopsis(cmdStr string) (Cmd, error) {
```
Comment (gongweibao) [inline go/file_manager/pfscommon/checkcmd.go]: Done

File: go/file_manager/pfscommon/cmd.go [hunk 4]
```diff
@@ -0,0 +1,18 @@
+package pfscommon
+
+/*
```
Comment (helinwang) [inline go/file_manager/pfscommon/cmd.go]: no commented out code in the code base, please.

File: go/file_manager/pfscommon/parsebody.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+	}
+}
+
+func Body2Json(w http.ResponseWriter,
+	r *http.Request,
+	req interface{},
+	rep Rep) error {
+	body, err := ioutil.ReadAll(io.LimitReader(r.Body, 2048))
+	if err != nil {
+		panic(err)
```
Comment (helinwang) [inline go/file_manager/pfscommon/parsebody.go]: This kind of utility function should return error, and caller need to do propery error handling (log error, retry...).
Please see following two blog posts for reference: https://blog.golang.org/error-handling-and-go , https://blog.golang.org/errors-are-values

File: go/file_manager/pfscommon/pfstrace.go [hunk 6]
```diff
@@ -0,0 +1,17 @@
+package pfscommon
+
+import (
+	"fmt"
+	"runtime"
+)
+
+func CallerFileLine() string {
+	_, fileName, fileLine, ok := runtime.Caller(2)
```
Comment (helinwang) [inline go/file_manager/pfscommon/pfstrace.go]: standard library provides file line logging: https://golang.org/pkg/log/#pkg-constants : " Lshortfile                    // final file name element and line number: d.go:23. overrides Llongfile"

File: go/file_manager/pfsmodules/getfiles.go [hunk 7]
```diff
@@ -0,0 +1,111 @@
+package pfsmodules
```
Comment (helinwang) [inline go/file_manager/pfsmodules/getfiles.go]: Need better package naming.
Package name could be `pfs`, and there is a struct called `Server`, so when people using it, can do:
```go
import ".../pfs"
var server pfs.Server
```
It's easier to understand.

Please see: https://blog.golang.org/package-names

File: go/file_manager/pfsmodules/storage_files.go [hunk 8]
```diff
@@ -0,0 +1 @@
+package pfsmodules
```
Comment (helinwang) [inline go/file_manager/pfsmodules/storage_files.go]: Please don't check in empty files :)

File: go/file_manager/pfsserver/handler.go [hunk 9]
```diff
@@ -0,0 +1,63 @@
+package pfsserver
+
+import (
+	"encoding/json"
+	"github.com/cloud/go/file_manager/pfscommon"
```
Comment (helinwang) [inline go/file_manager/pfsserver/handler.go]: Are there really benefit of having so much packages :) (pfscommon, pfsmodules), why not just have one package?

File: go/file_manager/pfsserver/main/main.go [hunk 10]
```diff
@@ -0,0 +1,23 @@
+package main
```
Comment (helinwang) [inline go/file_manager/pfsserver/main/main.go]: I think it's better to put main package files inside
`foo/cmd/cli/`
See https://github.com/coreos/etcd/tree/master/cmd for folder structure reference.

File: go/file_manager/pfscommon/parsebody.go [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+import (
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+)
+
+type Rep interface {
```
Comment (helinwang) [inline go/file_manager/pfscommon/parsebody.go]: I did not go through the whole code so I may be wrong:
We need to be careful not to overuse interface in golang,
Please see this post for some information: https://medium.com/@cep21/preemptive-interface-anti-pattern-in-go-54c18ac0668a

File: go/.gitignore [hunk 12]
```diff
@@ -0,0 +1 @@
+paddlecloud/paddlecloud
```
Comment (Yancey1989) [inline go/.gitignore]: [here](https://github.com/PaddlePaddle/cloud/blob/develop/go/paddlecloud/.gitignore#L1) is a `.gitignore` file.I think this is unnecessary.

File: go/file_manager/pfsmodules/command.go [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+
+const (
+	FileNotFound      = "no such file or directory"
+	DirectoryNotAFile = "should be a file not a directory"
+)
+
+type Command interface {
+	GetCmdAttr() *CmdAttr
+	GetResponse() Response
+	//PushRequest()
```
Comment (Yancey1989) [inline go/file_manager/pfsmodules/command.go]: We can delete the annotated code if it's unnecessary.

File: go/file_manager/pfscommon/checkcmd.go [hunk 14]
```diff
@@ -0,0 +1,37 @@
+package pfscommon
+
+/*
+type FilePath struct {
```
Comment (helinwang) [inline go/file_manager/pfscommon/checkcmd.go]: @gongweibao What is the purpose for the commented out code?

**Comment by helinwang** [comment]
This PR is getting really big. I think it would be good for review and progress if several smaller sized PRs are created, reviewed, and merged :)

**Comment by Yancey1989** [comment]
How about this PR? @gongweibao can you split this PR into some smaller one? such as a server framework, "ls" command and etc..

File: go/file_manager/pfsserver/main/main.go (hunk 10 above)
Comment (gongweibao) [inline go/file_manager/pfsserver/main/main.go]: Done

File: go/file_manager/pfsserver/handler.go (hunk 1 above)
Comment (gongweibao) [inline go/file_manager/pfsserver/handler.go]: Done

File: go/file_manager/pfscommon/parsebody.go (hunk 5 above)
Comment (gongweibao) [inline go/file_manager/pfscommon/parsebody.go]: Done

File: go/file_manager/pfscommon/parsebody.go [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+		log.Printf("encode err:%s\n", err.Error())
+		panic(err)
+	}
+}
+
+func Body2Json(w http.ResponseWriter,
+	r *http.Request,
+	req interface{},
+	rep Rep) error {
+	body, err := ioutil.ReadAll(io.LimitReader(r.Body, 2048))
```
Comment (gongweibao) [inline go/file_manager/pfscommon/parsebody.go]: Done

File: go/file_manager/pfscommon/parsebody.go (hunk 2 above)
Comment (gongweibao) [inline go/file_manager/pfscommon/parsebody.go]: Done

File: go/file_manager/pfscommon/parsebody.go (hunk 11 above)
Comment (gongweibao) [inline go/file_manager/pfscommon/parsebody.go]: Done

File: go/file_manager/pfsmodules/chunks.go [hunk 16]
```diff
@@ -0,0 +1,80 @@
+package pfsmodules
```
Comment (gongweibao) [inline go/file_manager/pfsmodules/chunks.go]: Done

File: go/file_manager/pfsserver/logger.go [hunk 17]
```diff
@@ -0,0 +1,23 @@
+package pfsserver
```
Comment (gongweibao) [inline go/file_manager/pfsserver/logger.go]: Done

File: go/file_manager/pfsserver/routes.go [hunk 18]
```diff
@@ -0,0 +1,27 @@
+package pfsserver
```
Comment (gongweibao) [inline go/file_manager/pfsserver/routes.go]: Done

File: go/file_manager/pfsserver/router.go [hunk 19]
```diff
@@ -0,0 +1,26 @@
+package pfsserver
```
Comment (gongweibao) [inline go/file_manager/pfsserver/router.go]: Done

File: go/file_manager/pfscommon/parsebody.go (hunk 5 above)
Comment (gongweibao) [inline go/file_manager/pfscommon/parsebody.go]: Done

File: go/file_manager/pfscommon/pfstrace.go (hunk 6 above)
Comment (gongweibao) [inline go/file_manager/pfscommon/pfstrace.go]: Done

File: go/file_manager/pfscommon/parsebody.go (hunk 11 above)
Comment (gongweibao) [inline go/file_manager/pfscommon/parsebody.go]: Done

File: go/file_manager/pfsserver/main/main.go (hunk 10 above)
Comment (gongweibao) [inline go/file_manager/pfsserver/main/main.go]: Done

File: go/file_manager/pfsserver/handler.go (hunk 9 above)
Comment (gongweibao) [inline go/file_manager/pfsserver/handler.go]: Done

File: go/file_manager/pfsmodules/storage_files.go (hunk 8 above)
Comment (gongweibao) [inline go/file_manager/pfsmodules/storage_files.go]: Done

File: go/.gitignore (hunk 12 above)
Comment (gongweibao) [inline go/.gitignore]: Done

File: go/file_manager/pfsmodules/command.go (hunk 13 above)
Comment (gongweibao) [inline go/file_manager/pfsmodules/command.go]: Done

File: go/file_manager/pfscommon/checkcmd.go (hunk 14 above)
Comment (gongweibao) [inline go/file_manager/pfscommon/checkcmd.go]: Done

File: go/file_manager/pfsmodules/getfiles.go (hunk 7 above)
Comment (gongweibao) [inline go/file_manager/pfsmodules/getfiles.go]: 多谢鹤林。我研究一下。

**Pull request closed without merging**